src/simulators/matrix_product_state/matrix_product_state_internal.cpp

```diff
@@ -173,6 +205,57 @@ uint_t reorder_qubits(const reg_t qubits, uint_t index) {
   return new_index;
 }
 
+template <class T>
+void permute_all_qubits(const std::vector<T>& orig_statevector, 
+			const reg_t &input_qubits,
+			const reg_t &output_qubits,
+			std::vector<T>& new_statevector) {
+  uint_t new_index;
+  uint_t length = 1ULL << input_qubits.size();   // length = pow(2, num_qubits)
+  // if qubits are [k0, k1,...,kn], move them to [0, 1, .. , n], but preserve relative
+  // ordering
+  reg_t squeezed_qubits(input_qubits.size());
+  squeeze_qubits(input_qubits, squeezed_qubits);
+
+  for (uint_t i=0; i < length; i++) {
+    new_index = permute_qubits(squeezed_qubits, i, output_qubits);
+    new_statevector[new_index] = orig_statevector[i];
+  } 
+}
+
+uint_t permute_qubits(const reg_t &input_qubits, uint_t index, const reg_t &output_qubits) {
+  uint_t new_index = 0;
+  // pos is from right to left, i.e., msb has pos 0, and lsb has pos num_qubits-1
+  // This is in line with the ordering of the qubits in the MPS structure
+  int_t current_pos = 0, current_val = 0, new_pos = 0, shift =0;
+  uint_t num_qubits = input_qubits.size();
+
+  for (uint_t in=0; in<num_qubits; in++) {
+    //current_pos = num_qubits-1-in;
+    current_pos = in;
+    for (uint_t out=0; out<num_qubits; out++) {
+      if (input_qubits[in] == output_qubits[out]) {
+	//new_pos = num_qubits-1-out;
+	new_pos = out;
+	break;
+      }
+    }
+    shift = new_pos - current_pos;
+    current_val = 1ULL << current_pos;
+
+    if (index & current_val) {
+      if (shift > 0) {
+	new_index += current_val << shift;
+      } else if (shift < 0) {
+	new_index += current_val >> -shift; 
+      } else {
+	new_index += current_val;
+      }      
+    }
+  }
+  return new_index;
```

Comment:
Consider rewriting `reverse_bits` and `reverse_all_bits`, such that they call new functions like `permute_qubits`, instead of reimplementing them for the special case of reversed bits.

Reply:
I wrote the comment above about rewriting `reorder_all_qubits` as a function of `permute_all_qubits`. Same goes for `reverse_all_bits`, although here I may decide to leave it, for simplicity reasons. Anyway, I will open a separate PR for this.